File: 29-divide-two-integers/29-divide-two-integers.cpp
class Solution {
public:
    int divide(int dividend, int divisor) {
        if (dividend == INT_MIN && divisor == -1)
            return INT_MAX;
        
        int sign = dividend > 0 ^ divisor > 0 ? -1 : 1;
        long dvd = labs(dividend), dvs = labs(divisor), ans = 0;
        
        while (dvs <= dvd)
        {
            long temp = dvs, m = 1;
            while ((temp << 1) <= dvd)
            {
                temp = temp << 1;
                m = m << 1;
            }
            ans += m;
            dvd -= temp;
        }
        return ans * sign;
    }
};

File: 29-divide-two-integers/29-divide-two-integers.py
class Solution:
    def divide(self, dividend: int, divisor: int) -> int:
        ans = 0
        factor = (-1) ** ((dividend < 0) ^ (divisor < 0))
        dividend, divisor = abs(dividend), abs(divisor)
        maxnum = 2147483647
        
        if divisor == 1:
            return min(factor * dividend, maxnum)
        
        while dividend >= divisor:
            
            dvd, dvs = dividend, divisor
            m = 1
            while dvd >= (dvs << 1):
                dvs <<= 1
                m <<= 1
            
            ans += m
            dividend -= dvs
            # print(dvs)
        
        return min(factor * ans, maxnum)
